### ISSUE TITLE
 About improvement of user customization experience.

### ISSUE THREAD

**Issue by misumisumi** [ayamir/nvimdots#888]
### Feature description

This is a suggestion regarding #885.
It contains major changes across the board.
Other suggestions welcome..

I'm looking for a way that doesn't change the original configuration too much.
Work is done at [misumisumi/nvimdots/add-user-custom-function](https://github.com/misumisumi/nvimdots/tree/add-user-custom-function).

- [x] Implementing the `load_plugin` function
  - [x] Extend default value when given value is a table
  - [x] Replace completely when the given value is a function
- [x] Support override for `event.lua` 
- [x] Support override for `options.lua` 
- [x] Support override for `settings.lua` 
- [x] Support override for `keymap.lua`

**Comment by misumisumi** [comment]
https://github.com/ayamir/nvimdots/pull/885#pullrequestreview-1542119001 ccording to A, an override that does not rely on LazyNvim is likely necessary.
-[] Consider using the merge function or better

**Comment by fecet** [comment]
Maybe we can take a look at https://github.com/LazyVim/LazyVim/tree/main, imitate its behavior or develop based on it.

**Comment by Jint-lzxy** [comment]
> Maybe we can take a look at https://github.com/LazyVim/LazyVim/tree/main, imitate its behavior or develop based on it.

Hmm I don't feel like things are gonna work this way. Cause compared to `LazyVim`, we used a completely different loading strategy & directory hierarchy (implemented in #458, that's why we [added a custom loader](https://github.com/ayamir/nvimdots/blob/main/lua/core/pack.lua) instead of using `lazy.nvim`'s). Following `LazyVim` will definitely require a complete rewrite - but (at least in the current state) we don't want this config to become some sort of "Neovim distribution", where the base config and user config are completely separated (or are connected via settings entries) - which means incremental updates to plugins' specs are almost impossible (well, this is the case if the user did follow the instructions to customize their experience). See https://github.com/ayamir/nvimdots/pull/885#pullrequestreview-1542119001 for some thoughts on this.

IMO this is not _that_ viable regarding our current situation, but we can further discuss its feasibility as well 😄

[Quoting](https://github.com/ayamir/nvimdots/issues/306#issuecomment-1293552965) @CharlesChiuGit:

> I once tried Lunarvim, but it also has it's own issue. One day u will still need to know how to fix bugs like this.
The donw side of a highly pre-config nvimdot is that u can't be sure about wether the bug is from lunarvim, plugins or nvim itself.

cc @fecet

**Comment by fecet** [comment]
Maybe we can discuss what's the best way to do custom config based on current setup. For me I [forked (https://github.com/fecet/nvim)this and do rebase from time to time，I don't really like this as I often have to deal with various conflicts.

If would be better IMO if adding or removing plugins are happen at files level (not tables), but I guess there is some better approach

**Comment by Jint-lzxy** [comment]
> If would be better IMO if adding or removing plugins are happen at files level (not tables), but I guess there is some better approach.

Yeah that's sort of what we're going to implement. Quoting https://github.com/ayamir/nvimdots/pull/885#pullrequestreview-1542119001:
> \<We could...\> wrap this as an API call like `require("...").load_plugin("<name-used-for-require>", current_spec)` where user configs will be automatically merged with parent specs, and if the passed user config is a function, it will completely replace its parent spec.

**Comment by fecet** [comment]
My point is, if we have to refactor the code anyway, we can choose to adopt the approach of lazyvim, which seems more user-friendly (for regular users) and easier to develop. 
One significant advantage of lazyvim I would like to mention is that it allows us to include some niche plugins as [extras](https://github.com/LazyVim/LazyVim/tree/main/lua/lazyvim/plugins/extras), and users can choose whether to import them or not.

I dont really understand why 
> incremental updates to plugins' specs are almost impossible

for neovim distributionn as I never use any of them, but lazyvim seems possible to do so?

**Comment by Jint-lzxy** [comment]
> My point is, if we have to refactor the code anyway, we can choose to adopt the approach of lazyvim, which seems more user-friendly (for regular users) and easier to develop and we can get free upstrem update(they appear to be very activte). Otherwise, I would rather keep the current configuration.

Yeah FWIW we are planning to use similar _logic_ compared with `LazyVim`. That is, "to separate user config from the base config to avoid merging conflicts (and users can get free upstrem update)". The thing is, our structure is different from `LazyVim`'s. `LazyVim` is somehow similar to our previous structure _('functionality-oriented', a file may contain multiple plugins and everything is placed there)_, while we currently use 'plugins-oriented' approach _(a file is treated as one basic unit for a single plugin, everything about that plugin is placed there)_ - configuring those plugins is roughly as follows:

### **If this is an existing plugin:**

Create a new file with the same name as the parent spec in `user/config`, and add the options that should be passed to the `setup()` function there. If the user want to completely replace the parent spec, simply pass a function instead.

#### Example:
- I want to modify the settings of `specs.nvim` (`ui/specs.lua`):

(Parent spec \[in the future(?)\]:)
```lua
return function()
	local parent_specs = {
		show_jumps = true,
		min_jump = 10,
		popup = {
			delay_ms = 0, -- delay before popup displays
			inc_ms = 10, -- time increments used for fade/resize effects
			blend = 10, -- starting blend, between 0-100 (fully transparent), see :h winblend
			width = 10,
			winhl = "PmenuSbar",
			fader = require("specs").pulse_fader,
			resizer = require("specs").shrink_resizer,
		},
		ignore_filetypes = {},
		ignore_buftypes = { nofile = true },
	}

	require("...").load_plugin("specs", parent_specs)
end
```
(Steps:)
1. Create `user/config/specs.lua`;

2.1. **I want to set `popup.delay_ms` to 20**:
```lua
return {
	popup = {
		delay_ms = 20,
	}
}
```
`load_plugin` will merge this spec with the parent spec, and then call `require("specs").setup({final_spec})`

2.2. **I want to customize this plugin myself**:
```lua
return function(opts) -- This is an empty table
	opts.show_jumps = true
	-- OR --
	opts = { show_jumps = true }
end
```
And `load_plugin` would instead do: `require("specs").setup({opts})`

### **If this is a new plugin:**
Similar to `LazyVim`:
> Adding a plugin is as simple as adding the plugin spec to one of the files under `user/plugins/*.lua`. You can create as many files there as you want.
>
> You can structure your lua/plugins folder with a file per plugin, or a separate file containing all the plugin specs for some functionality.
> ```lua
> return {
>   -- add symbols-outline
>   ["simrat39/symbols-outline.nvim"] = {
>     lazy = true,
>     cmd = "SymbolsOutline",
>     opts = {
>       -- add your options that should be passed to the setup() function here
>       position = "right",
>     },
>     -- OR --
>     config = function()
>      -- setup the plugin urself
>     end
>   },
> }
> ```

Does this make things a bit clearer?

**Comment by CharlesChiuGit** [comment]
> Yeah FWIW we are planning to use similar _logic_ compared with `LazyVim`. That is, "to separate user config from the base config to avoid merging conflicts (and users can get free upstrem update)". ....

I think this really nice, I'm buying it lol

**Comment by fecet** [comment]
Yes I like this way. But I'm wondering why our config is `plugins-oriented`,  doesn't our config also use a single file contains many configs, 
https://github.com/ayamir/nvimdots/blob/main/lua/modules/plugins/editor.lua looks very similar to https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/editor.lua and can be easily migrated from one to another one.

I think the correct way is use `functionality-oriented` for some complicated and basic part like treesitter, lsp, dap and telescope as they are hard to seperate and doesn't need much modified and use `plugins-oriented` for other plugins so they are easier to manager for users and developers.

**Comment by Jint-lzxy** [comment]
> I think the correct way is use `functionality-oriented` for some complicated and basic part like treesitter, lsp, dap and telescope as they are hard to seperate and doesn't need much modified and use `plugins-oriented` for other plugins so they are easier to manager for users and developers.

Sorry for the misleading wording - the main focus here is `configs`. `LazyVim` put the spec and its config together (making that a "comprehensive" file), but we separated them (each "config" file corresponds to exactly one _specification_ (main plugin + related dependencies)). The thing is, that (comprehensive file) makes examining/editing the change history in the version control system cumbersome. It may happen that the same file is being edited all the time and consequently more challenging to understand what and when was fixed (+possible merge conflicts). We chose this structure b/c most of the time we work with configs rather than plugin specifications 😄

**Comment by misumisumi** [comment]
@Jint-lzxy
Could you tell us a little more about the following  https://github.com/ayamir/nvimdots/pull/885#pullrequestreview-1542119001.
> this structure in https://github.com/ayamir/nvimdots/pull/458 is the opts key would completely replace parent specs when passed as a function.
---
@fecet
My PR is inspired by [LazyNvim](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-customizing-plugin-specs).
Some changes made by the user are as follows.
For example. override `alpha.lua`
step.1 `user/modules/plugins/ui.lua`
```
local ui = {}
ui["goolord/alpha-nvim"] = {
    enabled = false, # If you do not use plugin.
    opts = require("user.modules.configs.ui.alpha"),
}
return ui
```
step.2 `user/modules/configs/ui/alpha.lua`
```
return function()
	local alpha = require("alpha")
	local dashboard = require("alpha.themes.dashboard")
	require("modules.utils").gen_alpha_hl()

	dashboard.section.header.val = {
        [[ your ascii art ]],
	}
end
```
The current PR does not support full overrides.
You can pseudo-completely override by defining your own config.

**Comment by Jint-lzxy** [comment]
> Could you tell us a little more about the following https://github.com/ayamir/nvimdots/pull/885#pullrequestreview-1542119001.

The key point is [here (lazy.nvim/lua/lazy/core/plugin.lua)](https://github.com/folke/lazy.nvim/blob/3ad55ae678876516156cca2f361c51f7952a924b/lua/lazy/core/plugin.lua#L515-L534):
>```lua
> elseif type(values) == "function" then
>   ret = values(plugin, ret) or ret
>   return type(ret) == "table" and ret or { ret }
> end
>```

Because we are using [a custom loader](https://github.com/ayamir/nvimdots/blob/6de8afe0fc4a6e53a9d64a013327d5482b6dbe99/lua/core/pack.lua#L20-L56), the relative order in which plugins appear internally (inside `lazy.nvim`) is not specified. That is to say, it is possible that the parent spec appears after its children, which means this would block out any user configs. If we are sure to adopt this PR's structure, we need to rewrite the loader [_(using `lazy.nvim`'s preset)_](https://github.com/folke/lazy.nvim#%EF%B8%8F-importing-specs-config--opts).

**Comment by fecet** [comment]
Great! I feel like we are moving in the right direction. I still haven't find time to dig into lazyvim, is it possible to use our custom loader and import extras from lazyvim? So the config will look like
```lua
require("lazy").setup({
  spec = {
    -- add LazyVim and import its plugins
    { "ayamir/nvimdots", import = "nvim.plugins" },
    -- import any extras modules here
   { import = "lazyvim.plugins.extras.lang.typescript" },
    -- { import = "lazyvim.plugins.extras.lang.json" },
    -- { import = "lazyvim.plugins.extras.ui.mini-animate" },
    -- import/override with your plugins
    { import = "plugins" },
  },
```
I have many plugins that not suitable to add in main config but could be helpful for someone so it's good to have a `extra` part.

**Comment by Jint-lzxy** [comment]
> is it possible to use our custom loader and import extras from lazyvim

Yeah it's possible.

**Comment by misumisumi** [comment]
I thought of a function that would allow the user to override the plugin's options without making major changes to the current configuration.
If nil is given, it does the default table expansion if a table is given, or replaces it entirely if a function is given.
I think this works fine for many plugins.
However, like `wilder.nvim`, it doesn't work well if you set options other than `opts` after calling `setup`.
`load_plugin` must be called last.

- `load_plugin()
```lua
---@param plugin_name string @Plugin name
---@param opts table @Default options for plugin
---@param isLuaPlugin? boolean @If plugin is not made by lua
function M.load_plugin(plugin_name, opts, isLuaPlugin)
	isLuaPlugin = isLuaPlugin or true
	local ok, custom = pcall(require, "user.modules." .. plugin_name)
	if ok then
		if custom == nil then
			require(plugin_name).setup(false)
		elseif type(custom) == "table" then
			assert(
				isLuaPlugin,
				"This plugin is not made by lua, so define options in functions (probably using `vim.g.*`)"
			)
			opts = vim.tbl_deep_extend("force", opts, custom)
			require(plugin_name).setup(opts)
		elseif type(custom) == "function" then
			local user_opts = custom()
			if type(user_opts) == "table" and isLuaPlugin then
				require(plugin_name).setup(user_opts)
			end
		else
			error(
				"Please return `nil` if you disable plugin or `table` if you override config or `function` if you replace config completely "
			)
		end
	else
		require(plugin_name).setup(opts)
	end
end

```

**Comment by misumisumi** [comment]
You can isolate your own plugins by adding two lines to `pack.lua`.
If a user wants to include additional plugins, simply add them to `user/modules/plugins`.
```
	local get_plugins_list = function()
		local list = {}
		local plugins_list = vim.split(fn.glob(modules_dir .. "/plugins/*.lua"), "\n")
		local user_plugins_list = vim.split(fn.glob(modules_dir .. "/plugins/*.lua"), "\n")
		plugins_list = vim.list_extend(plugins_list, user_plugins_list)
```

**Comment by misumisumi** [comment]
IMO we have to think a bit about overriding keymap.
I simply implemented a method of overriding each item, but there is actually an order to load the keymap, so if you want to assign the key in another item to another command, you must include it in the file called later. plug.
Want, all item files are combined in `keymap/init.lua` to return a table. Then load it with neovim after overriding the user settings.
Separate settings for `lsp` and `gitsigns` and override them in functions.

**Comment by misumisumi** [comment]
I have implemented an override that does not significantly change the current file and directory structure. Could you please give me a review if possible?
Details are [misumisumi/nvimdots/add-user-custom-function](https://github.com/misumisumi/nvimdots/tree/add-user-custom-function).
An example of actually overriding using this is [misumisumi/nvimdots/my-config)](https://github.com/misumisumi/nvimdots/tree/my-config).

**Comment by Jint-lzxy** [comment]
> IMO we have to think a bit about overriding keymap.

In https://github.com/ayamir/nvimdots/commit/372d3c9f7dab8db15b5b7268442b2b3584801515 I introduced a `keymap` module that can help overwrite existing keymaps. Maybe something can be done there? Since we could use `user.keymap` to overwrite those base keymaps after they are loaded.

**Comment by misumisumi** [comment]
Implemented settings to isolate user overrides in newly created branches without major modifications to the current configuration.
Therefore, I closed the previous PR (#885) and created a new PR (#931).
